### DISCUSSION TITLE
Delete all versions of a secret

### DISCUSSION THREAD

**Discussion by sebv004** [ansible-collections/community.hashi_vault#392] [General]
Hi,
I'm trying to find a way to delete all versions of a secret.

I found community.hashi_vault.vault_kv2_delete module, but you can delete the latest version of the secret or you have to populate an array of the versions you want to delete only.

Is there an easy way to populate this array ?

Thank you in advance.

**Comment by briantist** [comment]
Hi @sebv004 , to get the list of versions for a given secret, you need to read the `metadata` path for the secret (which is not the same as the metadata key that's returned when reading the `data` path). Although we don't have a dedicated plugin for reading kv2 metadata, you can use the generic [`vault_read` lookup or module to do this](https://docs.ansible.com/ansible/devel/collections/community/hashi_vault/vault_read_module.html).

Have a look at the Vault API docs to see what the path and return structure is: https://developer.hashicorp.com/vault/api-docs/secret/kv/kv-v2#read-secret-metadata

If you were calling `vault_kv2_delete` with `mount_path=kv2` and `path=my/secret` then you would do your read against `kv2/metadata/my/secret`.

With the `data.versions` field you'll be able to loop over the versions and call the delete for any versions that are not already deleted/destroyed (you can check for a non-empty deletion time to determine if it's been deleted already).

I could probably come up with example Ansible for this but it might be a week or more before I'll be able to do that.

**Comment by AndrewSav** [comment]
Note that deleting all versions of a secret is not the same as deleting the secret. Even if you delete all the versions, the secret will still be there

**Reply by briantist** [reply]
Indeed, to delete an entire secret, we must delete the metadata, which isn't something we can do with this collection today:
https://developer.hashicorp.com/vault/api-docs/secret/kv/kv-v2#delete-metadata-and-all-versions

Thanks for submitting #399 @AndrewSav !
